contracts/EtomicSwapV2.sol

```diff
@@ -0,0 +1,483 @@
+        address taker,
+        bytes20 takerSecretHash,
+        bytes20 makerSecretHash,
+        uint32 paymentLockTime
+    ) external payable {
+        require(makerPayments[id].state == MakerPaymentState.Uninitialized, "Maker payment is already initialized");
+        require(taker != address(0), "Taker must not be zero address");
+        require(msg.value > 0, "ETH value must be greater than zero");
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                msg.value,
+                taker,
+                msg.sender,
+                takerSecretHash,
+                makerSecretHash,
+                address(0)
+            )
+        );
+
+        makerPayments[id] = MakerPayment(paymentHash, paymentLockTime, MakerPaymentState.PaymentSent);
+
+        emit MakerPaymentSent(id);
+    }
+
+    function erc20MakerPayment(
+        bytes32 id,
+        uint256 amount,
+        address tokenAddress,
+        address taker,
+        bytes20 takerSecretHash,
+        bytes20 makerSecretHash,
+        uint32 paymentLockTime
+    ) external {
+        require(makerPayments[id].state == MakerPaymentState.Uninitialized, "Maker payment is already initialized");
+        require(amount > 0, "Amount must not be zero");
+        require(taker != address(0), "Taker must not be zero address");
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                amount,
+                taker,
+                msg.sender,
+                takerSecretHash,
+                makerSecretHash,
+                tokenAddress
+            )
+        );
+
+        makerPayments[id] = MakerPayment(paymentHash, paymentLockTime, MakerPaymentState.PaymentSent);
+
+        emit MakerPaymentSent(id);
+
+        // Now performing the external interaction
+        IERC20 token = IERC20(tokenAddress);
+        // Ensure that the token transfer from the sender to the contract is successful
+        require(
+            token.transferFrom(msg.sender, address(this), amount),
+            "ERC20 transfer failed: Insufficient balance or allowance"
+        );
+    }
+
+    function spendMakerPayment(
+        bytes32 id,
+        uint256 amount,
+        address maker,
+        bytes20 takerSecretHash,
+        bytes32 makerSecret,
+        address tokenAddress
+    ) external {
+        require(makerPayments[id].state == MakerPaymentState.PaymentSent, "Invalid payment state. Must be PaymentSent");
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                amount,
+                msg.sender,
+                maker,
+                takerSecretHash,
+                ripemd160(abi.encodePacked(sha256(abi.encodePacked(makerSecret)))),
+                tokenAddress
+            )
+        );
+        require(paymentHash == makerPayments[id].paymentHash, "Invalid paymentHash");
+
+        makerPayments[id].state = MakerPaymentState.TakerSpent;
+
+        if (tokenAddress == address(0)) {
```

Comment:
I dont find event emission in `spendMakerPayment` function

Reply:
Will fix shortly :slightly_smiling_face: